### PR TITLE
In var foreach, use return type of the Current enumerator prop…

### PR DESCRIPTION
**Pull request by 333fred**
Fixes part of https://github.com/dotnet/roslyn/issues/39736. The deconstruction case is unhandled, and I'm going to leave it off until @jcouv gets back from vacation. It will require diving in to the tuple type, pulling it apart, and using the original type where appropriate. It would likely need to be rewritten after his changes, so we'll wait on doing it.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler please review.

**Comment by RikkiGibson** [comment]
>                             // foreach (var variable in collection)

looks like this comment is no longer applicable?

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:6927 in 53f1444. [](commit_id = 53f144454db31dbf4cc7962b7c7015ff770c94b9, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -6916,6 +6916,11 @@ public override void VisitForEachIterationVariables(BoundForEachStatement node)
                                 ReportNullabilityMismatchInAssignment(foreachSyntax.Type, sourceType, destinationType);
                             }
                         }
+                        else if (node.Syntax is ForEachStatementSyntax { Type: { IsVar: true } })
+                        {
+                            result = sourceState;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Is this redundant with the assignment on 6909?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -65221,6 +65221,53 @@ void M(string? s)
...
+    {
+        var items3 = new List<(int? Index, Alpha? Alpha)>()
+        {
+            (0, new Alpha { Value = ""A"" }),
+            (1, new Alpha { Value = ""B"" }),
+            (2, new Alpha { Value = ""C"" })
+        };
+        foreach (var indexAndAlpha in Identity(items3))
+        {
+            var(index, alpha) = indexAndAlpha;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: nit: formatting

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -65221,6 +65221,53 @@ void M(string? s)
...
+            index.ToString();
+            alpha.ToString(); // 1
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >alpha.ToString(); [](start = 12, length = 17)

Consider testing assignment to those variables, or displaying their inferred types with `VerifyTypes`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes.

---
In reply to: [345396562](https://github.com/dotnet/roslyn/pull/39769#discussion_r345396562) [](ancestors = 345396562)

**Pull request merged**